### PR TITLE
Refactor code base

### PR DESCRIPTION
**Pull request by nurmohammed840**

### CODE REVIEW

File: src/memory.rs [hunk 1]
```diff
@@ -157,63 +158,49 @@ pub struct ProcessMemoryStats {
...
+            cmd.truncate(50);
         }
-
-        self.pid = *pid;
-
-        let smaps_file = File::open(format!("/proc/{}/smaps", pid))?;
-        let mut reader = BufReader::new(smaps_file);
-        let mut line = String::new();
-
-        let mut mem_values = (0, 0, 0, 0);
```
Comment (xeome) [inline src/memory.rs:182]: This is here for data locality, you are reducing performance.

Comment (xeome) [reply src/memory.rs:182]: Also, bufreader is a more performant way to read files.

File: src/process.rs [hunk 2]
```diff
@@ -1,89 +1,61 @@
...
-                Some(task::spawn(async move { Process::new(pid) }))
-            })
-            .collect::<Vec<_>>();
-
-        // Wait for all the processes to finish
-        let mut processes = futures::future::try_join_all(processes).await?;
-        // Sort the processes by swap usage
-        processes.sort_by_key(|p| p.memory.swap);
-        self.processes = processes;
+    pub fn update(&mut self) -> io::Result<()> {
```
Comment (xeome) [inline src/process.rs:22]: Can you please consider the impact on performance while refactoring and find a way to maintain or improve performance without sacrificing functionality? Serializing this function leads to a major decrease in performance. Do not remove concurrency.

**Comment by xeome** [comment]
Thank you for your pull request. I have reviewed your changes, and while I appreciate your efforts, there are a few concerns that I would like to address before merging this code into the master branch.

Firstly, I noticed that you have removed the concurrency that was previously implemented, which has significantly reduced the performance of the code. Concurrency is an essential feature that helps achieve faster processing times. Therefore, I would request you to consider a way to reintroduce concurrency.

Secondly, I see that you have removed the temporary variables that were used for data locality, which has improved the CPU cache hit rate. Therefore, I would request you to find a way to reintroduce the temporary variables that were previously used for data locality to improve both cache hit rate and performance. 

Since this program is performance-critical, it is important to test for performance before merging any code changes into the master branch. Therefore, I kindly request that you run performance tests on your updated code changes to ensure that the code's performance has not been compromised in any way.

Your branch: 
`target/release/zmem  0.05s user 0.25s system 99% cpu 0.299 total`
Upstream master branch: 
`target/release/zmem  0.09s user 0.45s system 875% cpu 0.062 total`

Once again, thank you for your contributions, and I look forward to seeing your updates that address the concerns mentioned above.

Best regards,

File: src/memory.rs (hunk 1 above)
Comment (nurmohammed840) [reply src/memory.rs:182]: > bufreader is a more performant way to read files

It doesn't, `BufReader` is  only good when you have frequent small read.  

Also `fs::read_to_string` buffer the entire contents in memory.

File: src/process.rs (hunk 2 above)
Comment (nurmohammed840) [reply src/process.rs:22]: Operating system does't provide async API for handling file system.

> Then how does async runtime provide async API for file system ? 

It doesn't, tokio basically spawn a OS thread to handle  any blocking operation like filesystem. This is also same for any other async runtime like NodeJS.

From [tokio docs](https://tokio.rs/tokio/tutorial):

> Reading a lot of files. Although it seems like Tokio would be useful for projects that simply need to read a lot of files, Tokio provides no advantage here compared to an ordinary threadpool.

> Can you please consider the impact on performance while refactoring and find a way to maintain or improve performance without sacrificing functionality ?  
 
Initially I did use threads,  But my OS (with kernel v6.2)  didn't like it. With threads it took about 400 ms. without threads it took 440 ms to execute.

**Comment by nurmohammed840** [comment]
Well I have re-added  threads, But it didn't improve as much as I expected.

With threads took: 380ms +- 20ms
Without threads took: 470ms +- 10ms

> Secondly, I see that you have removed the temporary variables that were used for data locality, which has improved the CPU cache hit rate. Therefore, I would request you to find a way to reintroduce the temporary variables that were previously used for data locality to improve both cache hit rate and performance.

Sorry, I didn't understand this statement. Also what is you thought about  `BufReader`  for performance improvement ?

**Comment by xeome** [comment]
> > Secondly, I see that you have removed the temporary variables that were used for data locality, which has improved the CPU cache hit rate. Therefore, I would request you to find a way to reintroduce the temporary variables that were previously used for data locality to improve both cache hit rate and performance.

> Sorry, I didn't understand this statement. Also what is you thought about BufReader for performance improvement ?

Data locality refers to the concept of keeping data that is frequently accessed in close proximity to each other in memory. By using temporary variables for data locality, the code can ensure that frequently accessed data is kept close together in memory, improving the cache hit rate and overall performance. Removing these temporary variables can reduce cache hit rate and potentially decrease performance.

This is because modern CPUs have multiple levels of cache memory that can be accessed faster than main memory. When data is stored in cache, it can be accessed much more quickly, resulting in improved performance.

I don't know about bufreader, I guess I can try using normal `fs::read_to_string

> With threads took: 380ms +- 20ms
> Without threads took: 470ms +- 10ms

There's a big difference on my device, I'll ask my friends to test it too.

**Comment by xeome** [comment]
I apologize for the inconvenience, but after testing your branch, including the threaded version, it appears to be performing approximately 50% slower on both my machine and those of my friends. Can you try upstream master branch and compare timings?

**Pull request closed without merging**